webhooks/cluster.go

```diff
@@ -52,54 +58,191 @@ func (c *Cluster) Default(_ context.Context, obj runtime.Object) error {
 		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", obj))
 	}
 
-	// Note: The code in Default is intentionally not duplicated to avoid that we accidentally
-	// implement new checks in the API package and forget to duplicate them to the webhook package.
-	// The idea is to add new defaulting which requires a reader in the webhook package.
-	// When we drop the method in the API package we must inline it here.
-	cluster.Default()
+	if cluster.Spec.InfrastructureRef != nil && len(cluster.Spec.InfrastructureRef.Namespace) == 0 {
+		cluster.Spec.InfrastructureRef.Namespace = cluster.Namespace
+	}
+
+	if cluster.Spec.ControlPlaneRef != nil && len(cluster.Spec.ControlPlaneRef.Namespace) == 0 {
+		cluster.Spec.ControlPlaneRef.Namespace = cluster.Namespace
+	}
 
+	// If the Cluster uses a managed topology
+	if cluster.Spec.Topology != nil {
+		// tolerate version strings without a "v" prefix: prepend it if it's not there
+		if !strings.HasPrefix(cluster.Spec.Topology.Version, "v") {
+			cluster.Spec.Topology.Version = "v" + cluster.Spec.Topology.Version
+		}
+	}
 	return nil
 }
 
-// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+// ValidateCreate implements webhook.CustomValidator so a webhook will be registered for the type.
 func (c *Cluster) ValidateCreate(_ context.Context, obj runtime.Object) error {
 	cluster, ok := obj.(*clusterv1.Cluster)
 	if !ok {
 		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", obj))
 	}
-
-	// Note: The code in ValidateCreate is intentionally not duplicated to avoid that we accidentally
-	// implement new checks in the API package and forget to duplicate them to the webhook package.
-	// The idea is to add new validation which requires a reader and Cluster variable/patch validation
-	// in the webhook package.
-	// When we drop the method in the API package we must inline it here.
-	return cluster.ValidateCreate()
+	return c.validate(nil, cluster)
```

Comment:
nit: what about renaming the func receiver from `c` to `w` or to a more verbose `webhook`? (`c` tricked me into thinking it is an actual Cluster)

Reply:
same for ClusterClass

Reply:
webhook it is - I think you're right that there's a danger of misunderstanding becauase of the replicate Cluster{} struct here.